### ISSUE TITLE
RxSwift extensions question

### ISSUE THREAD

**Issue by kzaher** [ashfurrow/Moya#163]
Hi Ash,

I've seen this some time ago, but was swamped :)

I think there is a threading concern [here](https://github.com/ashfurrow/Moya/blob/master/Moya/RxSwift/Moya%2BRxSwift.swift#L20)
```
            if let existingObservable = self?.inflightRequests[endpoint] {
                return existingObservable
            }
```

We should probably use 
```
            objc_sync_enter(weakSelf)
            let inFlight = self?.inflightRequests[endpoint]
            objc_sync_exit(weakSelf)
            if let existingObservable = inFlight {
                return existingObservable
            }
```

It looks like ReactiveCocoa integration has the same [issue](https://github.com/ashfurrow/Moya/blob/master/Moya/ReactiveCocoa/Moya%2BReactiveCocoa.swift#L24)

If you want, I can make a PR, but just wanted to be sure this is a bug.

I think we are also [missing](https://github.com/ashfurrow/Moya/blob/master/Moya/RxSwift/Moya%2BRxSwift.swift#L38) sharing of a result construct (`>- variable`), I can add that also.

**Comment by alcarvalho** [comment]
Ok, so I have tried to fix this, but it broke my test when I added the `>- variable`. It won't complete whet this is present.

The sync is working well on RxSwift but it won't complete if added to the `ReactiveCocoaMoyaProvider`.

Here is the test for the RxSwift to see if you can help me with that. I don't like the way it is but I couldn't think of another way:

```swift
it("returns identical observables for inflight requests") {
    let target: GitHub = .Zen
    var response: MoyaResponse!
    let outerSignal = provider.request(target)
    outerSignal >- subscribeNext { (response) -> Void in
        expect(provider.inflightRequests.count).to(equal(1))
        let innerSignal = provider.request(target)
        innerSignal >- subscribeNext { (object) -> Void in
            expect(provider.inflightRequests.count).to(equal(1))
        }
    }
    expect(provider.inflightRequests.count).to(equal(0))
}
```

**Comment by ashfurrow** [comment]
Thanks for bringing up the issue, @kzaher – good catch! Hand't considered the signals coming from other threads >.<

@alcarvalho Thanks for the help! It's a tricky problem and a _very_ difficult use case to test. Could I ask you to open a pull request, with your work and new tests? We can have a conversation about it there :fireworks:

**Comment by alcarvalho** [comment]
Sure, @ashfurrow. Done! :wink:

**Comment by ashfurrow** [comment]
This was fixed when #166 was merged.
